geomstats/geometry/spd_matrices.py

```diff
@@ -711,12 +714,109 @@ def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
         inner_product : array-like, shape=[...,]
             Inner-product.
         """
-        spd_space = self.space
-        modified_tangent_vec_a =\
-            spd_space.inverse_differential_power(2, tangent_vec_a, base_point)
-        product = gs.einsum(
-            '...ij,...jk->...ik', modified_tangent_vec_a, tangent_vec_b)
-        result = gs.trace(product, axis1=-2, axis2=-1) / 2
+        eigvals, eigvecs = gs.linalg.eigh(base_point)
+        transp_eigvecs = gs.einsum('...ij->...ji', eigvecs)
+        rotated_tangent_vec_a = gs.matmul(transp_eigvecs, tangent_vec_a)
+        rotated_tangent_vec_a = gs.matmul(rotated_tangent_vec_a, eigvecs)
+        rotated_tangent_vec_b = gs.matmul(transp_eigvecs, tangent_vec_b)
+        rotated_tangent_vec_b = gs.matmul(rotated_tangent_vec_b, eigvecs)
+
+        ones = gs.ones(eigvals.shape)
+        vertical_index = gs.einsum('...i,...j->...ij', eigvals, ones)
+        horizontal_index = gs.einsum('...j,...i->...ij', eigvals, ones)
+        coefficients = 1 / (vertical_index + horizontal_index)
```

Comment:
The broadcasting rules allow to sum rows and columns to result in a matrix of pairwise sum, so instead of defining the horizontal and vertical indices you can simply use:
`coefficients = 1. / (eigvals[..., :, None] +  eigvals[..., None, :])`